### PR TITLE
Optional certbot and support for manual certificates

### PR DESCRIPTION
**Pull request by sanyud24**
issues:https:https://github.com/odk-x/tool-suite-X/issues/264
option to users to supply their own certificate

### CODE REVIEW

**Comment by wbrunette** [comment]
@sanyud24 great start you will also need to create a new docker-compose-https.yml different name and have nginx look for the manual certificate.

Basically simply remove everything but the nginx section (and the things nginx depends on like volumes) and rename the new file. 

The launch stack launch at the end will need to use the new file instead so you will need another if statement there.

**Comment by wbrunette** [comment]
@matt9j feel free to add any additional feedback.

File: init-odkx-sync-endpoint.py [hunk 1]
```diff
@@ -83,15 +83,19 @@ def run_interactive_config():
...
-          --email {} \
-          -d {} \
-          --rsa-key-size 4096 \
-          --agree-tos \
-          --cert-name bootstrap \
-          --keep-until-expiring \
-          --non-interactive".format(email, domain))
+        
+        manual_certificate = print("Do you wish to supply your own ssl certificate? [(Y)/n]")
+        if manual_certificate== "":
```
Comment (matt9j) [inline init-odkx-sync-endpoint.py]: I don't believe this python code is not correct. Right now if the user enters n, no certificate will be generated. I may be misunderstanding, but I think this line should be something along the lines of `if proceed.strip().lower()[0] != "y":`, assuming you accept my suggestion that the default should be "n"

File: init-odkx-sync-endpoint.py [hunk 2]
```diff
@@ -83,15 +83,19 @@ def run_interactive_config():
...
-        os.system("sudo certbot certonly --standalone \
-          --email {} \
-          -d {} \
-          --rsa-key-size 4096 \
-          --agree-tos \
-          --cert-name bootstrap \
-          --keep-until-expiring \
-          --non-interactive".format(email, domain))
+        
+        manual_certificate = print("Do you wish to supply your own ssl certificate? [(Y)/n]")
```
Comment (matt9j) [inline init-odkx-sync-endpoint.py]: I would suggest that the default behavior should be to automatically generate a certificate. Users having their own certificate is a special case. I'll defer to @wbrunette 's preference though.

File: init-odkx-sync-endpoint.py [hunk 3]
```diff
@@ -83,15 +83,19 @@ def run_interactive_config():
...
+        if manual_certificate== "":
+            manual_certificate = "n"
+
+                os.system("sudo certbot certonly --standalone \
```
Comment (matt9j) [inline init-odkx-sync-endpoint.py]: This entire function call should be one indent level less. In python the indentation is important and semantically part of the code.

File: init-odkx-sync-endpoint.py [hunk 4]
```diff
@@ -83,15 +83,19 @@ def run_interactive_config():
...
-          -d {} \
-          --rsa-key-size 4096 \
-          --agree-tos \
-          --cert-name bootstrap \
-          --keep-until-expiring \
-          --non-interactive".format(email, domain))
+        
+        manual_certificate = print("Do you wish to supply your own ssl certificate? [(Y)/n]")
+        if manual_certificate== "":
+            manual_certificate = "n"
```
Comment (matt9j) [inline init-odkx-sync-endpoint.py]: Where is this assignment used?

File: docker-compose-https_nginx.yml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    # Reload nginx every 12 hours to poll for new certificate
+    # refreshes which happen in the background.
+    command: /bin/bash -c "\
+      while :; do \
+        sleep 12h & wait $${!}; \
+        nginx -s reload; \
+      done & \
+      nginx -g \"daemon off;\" \
+      "
+    volumes:
```
Comment (matt9j) [inline docker-compose-https_nginx.yml]: How is the user-provided certificate added to nginx? Or is that not part of this pull request?

**Review by matt9j — CHANGES_REQUESTED** [review]
Hi @sanyud24, thanks for this contribution! I have a couple of questions opened in a review-- let me know if I can help clarify anything. Also more generally, were you able to test these changes? It is usually a good practice to test changes before submitting them, and add to your pull request how you did the testing. That can help me as a reviewer reproduce how things are supposed to work!

**Comment by sanyud24** [comment]
Really sorry to trouble you @matt9j  , I opened this  pullrequest to know if I'm on the right file , doing the right changes , I have my exams going on for a week now I'll correct everything in 2 -3 days after testing all the changes , I made some of the changes you suggested above a while ago but didn't commited them, I'll  commit after full testing .  Thankyou for the review.

**Comment by matt9j** [comment]
> Really sorry to trouble you @matt9j , I opened this pullrequest to know if I'm on the right file , doing the right changes , I have my exams going on for a week now I'll correct everything in 2 -3 days after testing all the changes , I made some of the changes you suggested above a while ago but didn't commited them, I'll commit after full testing . Thankyou for the review.

No problem! Good luck with your exams, and I'm looking forward to seeing the finished product : ) You are indeed on the right track!

**Comment by sanyud24** [comment]
I'm really not getting how to connect nginx code with dockerfile after storing private key and ssl certificate path . I could not test because of that , I'm sorry but I'm not getting anything to proceed with after this.

File: init-odkx-sync-endpoint.py (hunk 2 above)
Comment (sanyud24) [reply init-odkx-sync-endpoint.py]: Fixed now, default is N

File: init-odkx-sync-endpoint.py (hunk 3 above)
Comment (sanyud24) [reply init-odkx-sync-endpoint.py]: yeah, missed it.. thanks for pointing it out

File: docker-compose-https_nginx.yml (hunk 5 above)
Comment (sanyud24) [reply docker-compose-https_nginx.yml]: added now

**Comment by sanyud24** [comment]
@matt9j @wbrunette Fixed the changes as per the suggestions and with correct logic.
Please review the PR.

I did test the script with both manual script and using certbot, but since I didn't had an actual domain to point to, I couldn't test the nginx container on my own. But the configs and the volume mounting of custom certificates is tested and working.

**Comment by wbrunette** [comment]
@matt9j for some reason it's not letting me request a re-review from you so consider this comment my request :)

File: init-odkx-sync-endpoint.py [hunk 6]
```diff
@@ -48,65 +49,79 @@ def run_interactive_config():
 
     while True:
         print("Would you like to enforce HTTPS? We recommend yes.")
-        enforce_https = input("enforce https [(Y)/n]:")
+        enforce_https = input("enforce https [(Y)/n]:").lower().strip()
         if enforce_https == "":
             enforce_https = "y"
-            enforce_https = enforce_https.lower().strip()[0]
+            enforce_https = enforce_https[0]
```
Comment (matt9j) [inline init-odkx-sync-endpoint.py:55]: Why the index access here? I think it should be on line 52 instead? The intention I think was to allow someone to type "y" or "yes".

File: docker-compose-https.yml [hunk 7]
```diff
@@ -1,112 +1,12 @@
...
-    internal: false
-
-volumes:
-  # A shared volume to allow the certbot to publish challenge
-  # responses which will be served by nginx.
-  www-certbot-vol:
-  # A shared volume to hold automatically issued certs and share them
-  # between certbot and nginx.
-  https-certs-vol:
+      - ${CERT_FULLCHAIN_PATH}:/etc/nginx/certs/fullchain.pem
```
Comment (matt9j) [inline docker-compose-https.yml:6]: Do you know if these environment variables remain set if the system is rebooted and the swarm has to restart? I see they are set in the launch command in the python script, but I don't know if they'll be captured automatically by docker swarm.

**Review by matt9j — COMMENTED** [review]
Generally this looks good to me. There is one issue with the environment variables you might want to double check. I don't have a test environment setup for this right now to test myself, but wanted to get some feedback to you without waiting another day :)

File: init-odkx-sync-endpoint.py (hunk 6 above)
Comment (sanyud24) [reply init-odkx-sync-endpoint.py:55]: @matt9j yeah, the logic is kinda broken when the string is non-empty but not y or n, i.e 'yes' or 'no'. It'll keep retrying since the loop won't exit.
However, I didn't change the logic so it's out of scope of this PR. Would you still like me to fix it here or in another PR?

File: docker-compose-https.yml (hunk 7 above)
Comment (sanyud24) [reply docker-compose-https.yml:6]: Yeah, it should remain persisted since they are present in the generated compose config for the docker swarm.

**Review by matt9j — APPROVED** [review]
Looks good to me per discussion in previous review.

File: init-odkx-sync-endpoint.py (hunk 6 above)
Comment (matt9j) [reply init-odkx-sync-endpoint.py:55]: Leaving it here is okay.

**Review by wbrunette — APPROVED** [review]
